server/server.ts

```diff
@@ -0,0 +1,191 @@
+  if (userExists) {
+    res.status(400).json({ message: 'User already exists' })
+    return
+  }
+
+  // Hash the password
+  const hashedPassword = bcrypt.hashSync(password, 10)
+
+  // Create a new user object
+  const newDbUser: DbUser = {
+    id: dbUsers.length + 1,
+    firstName,
+    lastName,
+    email,
+    gender,
+    role: 'User',
+    password: hashedPassword,
+  }
+
+  // Save the new user to the "users" collection in the database
+  router.db.get('users').push(newDbUser).write()
+
+  res.status(201).json({
+    message: 'User registered successfully',
+    userDetails: convertDbUserToUser(newDbUser),
+  })
+})
+
+interface LoginRequestBody {
+  email: string
+  password: string
+}
+
+// Set up a custom route for /login
+server.post('/login', (req, res) => {
+  const { email, password }: LoginRequestBody = req.body
+
+  if (!email || !password) {
+    res.status(400).json({ message: 'Email and password are required' })
+    return
+  }
+
+  // Check if user exists
+  const users = router.db.get('users').value()
+  const dbUser = users.find((user) => user.email === email)
+
+  if (!dbUser) {
+    res.status(400).json({ message: 'User not found' })
+    return
+  }
+
+  // Verify the password
+  const passwordIsValid = bcrypt.compareSync(password, dbUser.password)
+  if (!passwordIsValid) {
+    res.status(400).json({ message: 'Invalid password' })
+    return
+  }
+
+  // Create a JWT token
+  const token = jwt.sign(
+    { userId: dbUser.id, email: dbUser.email, role: dbUser.role },
+    JwtSecretKey,
+    {
+      expiresIn: JwtExpirationTime,
+    },
+  )
+
+  res.status(200).json({ message: 'Login successful', token })
+})
+
+interface RequestWithUser extends Request {
+  user: User
+}
+
+const authenticateJwtMiddleware = (
+  req: Request,
+  res: Response,
+  next: NextFunction,
+) => {
+  const token = req.headers['authorization']?.split(' ')[1] // Extract token from 'Authorization' header
+
+  if (!token) {
+    res.status(401).json({ message: 'Invalid or expired token' }) // Forbidden if no token is found
+    return
+  }
+
+  jwt.verify(token, JwtSecretKey, (err, user) => {
+    if (err) {
+      res.status(401).json({ message: 'Invalid or expired token' }) // Forbidden if token is invalid or expired
+      return
+    }
+
+    ;(req as RequestWithUser).user = user as User // Attach the user information (payload) to the request object
+    next() // Call the next middleware or route handler
+  })
+}
+
+server.get('/users', authenticateJwtMiddleware, (req, res) => {
+  const user = (req as RequestWithUser).user
+
+  if (user.role !== 'Admin') {
+    res.status(403).json({ message: 'Forbidden' })
+    return
+  }
+
+  const users = router.db.get('users').value()
+  res.status(200).json({ users })
+})
+
+// Set up middlewares and router
+server.use(middlewares)
+server.use(router)
+
+// Start server on a specific port
+const port = process.env.PORT || 3000
```

Comment:
As a suggestion, to keep a centralized config and make it more clear and readable, let's use a config to hold all information, create it at the top of the file, or in a separate file.

```
const config = {
  jwtSecret: process.env.JWT_SECRET_KEY || 'default-secret-key',
  jwtExpiration: process.env.JWT_EXPIRATION_TIME || '1h',
  port: process.env.PORT || 3000,
  dbPath: './db.json'
}
```